transports/janus_mqtt.c

```diff
@@ -0,0 +1,648 @@
+	}
+}
+
+void janus_mqtt_client_subscribe_failure(void *context, MQTTAsync_failureData *response) {
+	janus_mqtt_context *ctx = (janus_mqtt_context *)context;
+	int rc = response ? response->code : 0;
+	JANUS_LOG(LOG_ERR, "MQTT client has been failed subscribing to MQTT topic: %s, return code: %d. Reconnecting...\n", ctx->subscribe.topic, rc);
+
+	{ // reconnect
+		int rc = janus_mqtt_client_reconnect(context);
+		if(rc != MQTTASYNC_SUCCESS) {
+			JANUS_LOG(LOG_FATAL, "Can't reconnect to MQTT broker, return code: %d\n", rc);
+		}
+	}
+}
+
+int janus_mqtt_client_admin_subscribe(janus_mqtt_context *ctx) {
+	MQTTAsync_responseOptions options = MQTTAsync_responseOptions_initializer;
+	options.onSuccess = janus_mqtt_client_admin_subscribe_success;
+	options.onFailure = janus_mqtt_client_admin_subscribe_failure;
+	options.context = ctx;
+	return MQTTAsync_subscribe(ctx->client, ctx->admin.subscribe.topic, ctx->admin.subscribe.qos, &options);
+}
+
+void janus_mqtt_client_admin_subscribe_success(void *context, MQTTAsync_successData *response) {
+	janus_mqtt_context *ctx = (janus_mqtt_context *)context;
+	JANUS_LOG(LOG_INFO, "MQTT client has been successfully subscribed to MQTT topic: %s\n", ctx->admin.subscribe.topic);
+}
+
+void janus_mqtt_client_admin_subscribe_failure(void *context, MQTTAsync_failureData *response) {
+	janus_mqtt_context *ctx = (janus_mqtt_context *)context;
+	int rc = response ? response->code : 0;
+	JANUS_LOG(LOG_ERR, "MQTT client has been failed subscribing to MQTT topic: %s, return code: %d. Reconnecting...\n", ctx->admin.subscribe.topic, rc);
+
+	{ // reconnect
+		int rc = janus_mqtt_client_reconnect(context);
+		if(rc != MQTTASYNC_SUCCESS) {
+			JANUS_LOG(LOG_FATAL, "Can't reconnect to MQTT broker, return code: %d\n", rc);
+		}
+	}
+}
+
+int janus_mqtt_client_publish_janus_message(janus_mqtt_context *ctx, char *payload) {
+	MQTTAsync_message msg = MQTTAsync_message_initializer;
+	msg.payload = payload;
+	msg.payloadlen = strlen(payload);
+	msg.qos = ctx->publish.qos;
+	msg.retained = 0;
+
+	MQTTAsync_responseOptions options;
+	options.onSuccess = janus_mqtt_client_publish_janus_success;
+	options.onFailure = janus_mqtt_client_publish_janus_failure;
+	options.context = ctx;
+	return MQTTAsync_sendMessage(ctx->client, ctx->publish.topic, &msg, &options);
+}
+
+void janus_mqtt_client_publish_janus_success(void *context, MQTTAsync_successData *response) {
+	janus_mqtt_context *ctx = (janus_mqtt_context *)context;
+	JANUS_LOG(LOG_HUGE, "MQTT client has been successfully published to MQTT topic: %s\n", ctx->publish.topic);
+}
+
+void janus_mqtt_client_publish_janus_failure(void *context, MQTTAsync_failureData *response) {
+	janus_mqtt_context *ctx = (janus_mqtt_context *)context;
+	int rc = response ? response->code : 0;
+	JANUS_LOG(LOG_ERR, "MQTT client has been failed publishing to MQTT topic: %s, return code: %d\n", ctx->publish.topic, rc);
+}
+
+int janus_mqtt_client_publish_admin_message(janus_mqtt_context *ctx, char *payload) {
```

Comment:
Can't some of this methods, which are basically the same code, just envisage something like an additional `gboolean admin` flag, and then use the Janus or Admin stuff according to its value? I guess it makes sense for some callbacks to remain separated, especially if you don't have anything they tell you to allow you to retrieve the original context, for active methods should allow for that.